File: apps/api/routes/shorten.py
import uuid
from fastapi import APIRouter, Depends, Request
from sqlalchemy.ext.asyncio import AsyncSession
from nanoid import generate

from database import get_db
from models import Link
from schemas import ShortenRequest, ShortenResponse
from redis_client import redis, short_code_key, REDIS_TTL_SHORT_CODE
from config import config
from limiter import limiter

router = APIRouter()


@router.post("/shorten", response_model=ShortenResponse, status_code=201)
@limiter.limit("20/minute")
async def shorten(request: Request, body: ShortenRequest, db: AsyncSession = Depends(get_db)):
    url = body.url
    code = generate(size=config.short_code_length)

    link = Link(id=str(uuid.uuid4()), code=code, original_url=url)
    db.add(link)
    await db.commit()
    await db.refresh(link)

    await redis.set(short_code_key(code), url, ex=REDIS_TTL_SHORT_CODE)

    return ShortenResponse(
        code=link.code,
        short_url=f"{config.base_url}/{link.code}",
        original_url=link.original_url,
        created_at=link.created_at.isoformat(),
    )

File: apps/api/src/routes/shorten.ts
import { Router } from "express";
import { z } from "zod";
import { nanoid } from "nanoid";
import { prisma } from "../lib/prisma.js";
import { redis, REDIS_KEYS, REDIS_TTL } from "../lib/redis.js";
import { config } from "../lib/config.js";
import type { ShortenResponse } from "@url-shortener/shared";

const bodySchema = z.object({
  url: z.string().url({ message: "Must be a valid URL" }).max(2048),
});

export const shortenRouter = Router();

shortenRouter.post("/shorten", async (req, res) => {
  const parsed = bodySchema.safeParse(req.body);
  if (!parsed.success) {
    return res.status(400).json({
      error: "INVALID_URL",
      message: parsed.error.errors[0].message,
    });
  }

  const { url } = parsed.data;
  const code = nanoid(config.shortCode.length);

  const link = await prisma.link.create({
    data: { code, originalUrl: url },
  });

  await redis.set(REDIS_KEYS.shortCode(code), url, "EX", REDIS_TTL.shortCode);

  const response: ShortenResponse = {
    code: link.code,
    shortUrl: `${config.baseUrl}/${link.code}`,
    originalUrl: link.originalUrl,
    createdAt: link.createdAt.toISOString(),
  };

  return res.status(201).json(response);
});
